docs/adr/00004-add-ast-path-info-to-visitor.md

```diff
@@ -0,0 +1,47 @@
+# Provide more information to visitor methods
+
+-   Status: accepted <!-- optional -->
+-   Deciders: @kdy1, @kwonoj <!-- optional -->
+-   Date: 2022-06-30 <!-- optional -->
+
+## Context and Problem Statement
+
+Currently one can't get any information about parent node.
+This is due to ownership restriction of rust.
+We can't pass a reference to the parent to a visitor method.
+Let's say we have a `CallExpr`, and we are now going to call `visit_mut_callee` from `visit_mut_call_expr`.
+We want `&mut callee`, because it's the signature of `visit_mut_callee`.
+So we **borrow it from CallExpr**.
+As we mutabily borrowed something from `CallExpr`, we cannot pass `&CallExpr` to `visit_mut_callee`.
+
+We can workaround this by
+
+-   taking `CallExpr.callee`
+-   call `visit_mut_callee(&mut self, callee: &mut Callee, parent: &CallExpr)`, with dummy callee in `parent.callee`.
+-   restore the original callee from `visit_mut_call`.
+
+Doing this by hand is error-prone and doing this automatically by codegen is costly.
+This is explicit `memmove`, and `memmove` is quite costly.
+SWC moved from `Fold` to `VisitMut` because of `mmemove`.
+
+## Decision Drivers
+
+-   Help plugin authors writing plugins.
+
+## Considered Options
+
+-   [option 1] Stay with as-is.
+
+-   [option 2] Provide full information using `unsafe`.
+
+-   **[option 3] Provide small amount of information .**
```

Comment:
It 'd be good if we could list a few of what kind of things will be exposed with this approach.

Reply:
👍 
I'll add it tomorrow as I'm going to sleep

Reply:
Sure thing, no hurries. Just came to think of one more thing :

- what we won't try to expose as it is fundamentally related to the constraint explained in context, like mutable parent node itself or similar.